Linux and GCC support

Zakhrov opened the pull request:
I have cleaned up the Vulkan backend to get it to compile on Linux with GCC 9.1 and replaced the DirectXMath library with GLM. The code compiles successfully and creates the static libraries but I haven't been able to run the [glTFSample](https://github.com/GPUOpen-LibrariesAndSDKs/glTFSample) project

YvanDaSilva commented:
I don't have time to try it on Windows again, did you try it ?

I just had a quick look at the changes and at the code quality of this repo in general.
Has this sdk been reviewed by someone before ?
Lots of random spaces, lack of name consistency, src contains a different license file and other "basic" workflow stuff that shouldn't be commit like that.

I would suggest that a cleanup occur before this PR is applied, so the change is a bit clearer, and maybe travis is also added so we can ensure Cross platform compatibility at least at build time.
Maybe lets get some info about other gpuopen repos like:
https://github.com/GPUOpen-LibrariesAndSDKs/Anvil
https://github.com/GPUOpen-LibrariesAndSDKs/VulkanMemoryAllocator

luntik2012 commented:
please, add `build $ cmake .. -DGFX_API=VK/DX12` to build instructions in readme.md,

Zakhrov commented:
> please, add `build $ cmake .. -DGFX_API=VK/DX12` to build instructions in readme.md,

Done in the latest commit on my fork

janisozaur commented on src/VK/CMakeLists.txt:
```diff
@@ -1,6 +1,10 @@
+if(MSVC)
+  add_compile_options(/MP)
+else()
+add_compile_options(-M)
```
Is it really what you intended? I wouldn't expect generating make rules here.

Mixaill replied:
There was `add_compile_options(/MP)` in the `src/DX12` which is suprising too.

Zakhrov replied:
`add_compile_options(/MP)` gives me a build error in GCC 9
Passing the `-M` switch as a compile option seems to resolve it

Mixaill commented on src/common/Misc/WirePrimitives.cpp:
```diff
@@ -17,10 +17,10 @@
 #include "WirePrimitives.h"
-#include <DirectXMath.h>
-using namespace DirectX;
+#include <glm/glm.hpp>
```
Both **DirectXMath** and **GLM** are not needed here.
It is enough to delete `include` and `using namespace` here

Mixaill commented on readme.md:
```diff
@@ -62,20 +62,30 @@ Note: more info on the rendering backends can be found in the Readme of their re
+## Linux Prerequisites
+
+- [CMake 3.4](https://cmake.org/download/)
+- [GLM](https://glm.g-truc.net/0.9.9/index.html)
```
Unnecessary

Mixaill commented:
@Zakhrov I've updated a bit your branch with further fixes. Now Cauldron compiles on Windows and Linux, but there is a big amount of work to get it into the working state.

https://github.com/Mixaill/Cauldron/tree/linux-fixes

Zakhrov commented:
> @Zakhrov I've updated a bit your branch with further fixes. Now Cauldron compiles on Windows and Linux, but there is a big amount of work to get it into the working state.
> 
> https://github.com/Mixaill/Cauldron/tree/linux-fixes
I still had problems resolving <DirectXMath.h> It seems to be working now that I passed in a relative include path. I still have to see if the glTFSample compiles

aguaviva commented:
Maintainer here, just to set some expectations, the plan is to not support linux officially (maybe for now) but I still pay attention to all your comments and suggestions, so from now on I will try to be more GCC friendly :)

Zakhrov commented:
> Maintainer here, just to set some expectations, the plan is to not support linux officially (maybe for now) but I still pay attention to all your comments and suggestions, so from now on I will try to be more GCC friendly :)

OK thanks for letting us know. Will close the PR for now

Hi-Angel commented:
For the sake of cross-references: this fixes https://github.com/GPUOpen-LibrariesAndSDKs/Cauldron/issues/1

Zakhrov commented:
I've updated the code to Cauldron V1.4.1 and fixed the build errors with GCC. There is still a lot of work to be done before its actually usable though.

TODO
======

-  Create Linux specific Vulkan surfaces using VkXlibSurfaceCreateInfoKHR and VkWaylandSurfaceCreateInfoKHR
- Port sample glTF projects to Linux

Some doubts
==========
- Can we move the Vulkan components from DirectXMath to GLM? I'm right now using a patched fork of DirectXMath that compiles without the sal.h headers that are Windows specific. I don't think that would be feasible in the long run.
- It looks like it uses the DirectX shader compiler for compiling shaders on both DX12 and Vulkan. Can we use the default SPV shader compiler along with GLSL vertex and fragment shaders on Vulkan?

Mixaill commented:
>Can we move the Vulkan components from DirectXMath to GLM? I'm right now using a patched fork of DirectXMath that compiles without the sal.h headers that are Windows specific. I don't think that would be feasible in the long run.

* You should just use the vanilla DirectXMath and MIT-licensed SAL from https://github.com/dotnet/corert/blob/master/src/Native/inc/unix/sal.h
* At the first look there are only few usages of DirectXMath (Vec3/Vec4 and several constants), the other math was converted to the Sony vectormath library. I think it should be easy to eliminate DirectXMath usage for non-D3D code at least.

> It looks like it uses the DirectX shader compiler for compiling shaders on both DX12 and Vulkan. Can we use the default SPV shader compiler along with GLSL vertex and fragment shaders on Vulkan?

We can switch from the Windows SDK provided compiler to the github version https://github.com/Microsoft/DirectXShaderCompiler

Micket commented on src/common/Misc/Misc.cpp:
```diff
@@ -121,8 +134,6 @@ void Trace(const char* pFormat, ...)
-    FILE *file;
-
     //Open file
     if (fopen_s(&file, name, isbinary ? "rb" : "r") != 0)
```
This looks unintentional? `file` will be undefined below here, this can't be right.

Micket reviewed (commented):
Disclaimer: I'm not affiliated with AMD in any way, I just wanted to give some feedback to hope that this could move forward.
I have some specific comments, but, overall, I believe a PR like this is completely unmergeable. It's not even review-able. I'm going to echo @YvanDaSilva here, but it's just waaaaay to big, with waaaay to many changes unrelated to the specific purpose.

1. Whitespace changes: Either leave as is or fix in it's own PR.
2. nullptr modernization: Either leave as is or fix in it's own PR.
3. UINT8 -> uint8_t etc.: Either leave (just introduce define flags for them on linux if necessary) or replace in it's own PR.
4. Fix case errors with filenames (without linux changes) as much as possible in separate PR
and so on (probably splitting up the remaining
then the rest of the interesting fixes can be added gradually, preferably in a few distinct PRs working ones way up to a fully working build. 

I would create a new branch, cherry pick the changes from here and open up PRs comparable to the steps above (or according to whatever format AMD prefers) to break this down
Then, possibly rebase what is left of this PR on that.

Again: I don't speak for anyone else but myself here.

simsilver commented on CMakeLists.txt:
```diff
@@ -1,5 +1,11 @@
+set(CMAKE_GENERATOR_PLATFORM x64)
+add_compile_options(/MP)
+else()
+add_compile_options(-M)
```
this hide all compile error
